design/metadata-handling.md

```diff
@@ -0,0 +1,456 @@
+reference containing the name and the namespace of the secret containing the
+network configuration for this Metal3Machine.
+
+A `dataTemplate` field would be added, consisting of an object reference
+to a Metal3DataTemplate object containing the templates for
+the metadata and network data generation for this Metal3Machine.
+A `renderedData` field will be added in the status and will be a reference to
+the Metal3Data object created for this machine. If the dataTemplate field is set
+but either the `renderedData`, `metaData` or `networkData` fields in the status
+are unset, then the Metal3Machine controller will wait until it can find the
+Metal3Data object and the rendered secrets. It will then populate those fields.
+
+The secret containing the metaData or the network data could be provided by the
+user directly using the `metaData` or `networkData` fields in the spec of the
+object.
+
+When CAPM3 controller will set the different fields in the BareMetalHost,
+it will reference the metadata secret and the network data secret
+in the BareMetalHost. If any of the `metaData` or `networkData` status fields
+are unset, that field will also remain unset on the BareMetalHost.
+
+When the Metal3Machine gets deleted, the CAPM3 controller will remove its
+ownerreference from the data template object. This will trigger the deletion of
+the generated Metal3Data object and the secrets generated for this machine.
+
+A new object would be created, a Metal3DataTemplate type.
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3DataTemplate
+metadata:
+  name: nodepool-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    controller: true
+    kind: Metal3Cluster
+    name: cluster-1
+spec:
+  metaData: |
+    abc: def
+    local-hostname: {{ machineName }}
+    index: {{ index }}
+  networkData: |
+    {
+        "links": [
+            {
+                "id": "enp1s0",
+                "type": "phy",
+                "ethernet_mac_address": "{{ bareMetalHostMACByName "eth0" }}"
+            },
+            {
+                "id": "enp2s0",
+                "type": "phy",
+                "ethernet_mac_address": "{{ bareMetalHostMACByName "eth1" }}"
+            }
+        ],
+        "networks": [
+            {
+                "id": "Provisioning",
+                "type": "ipv4_dhcp",
+                "link": "enp1s0"
+            },
+            {
+                "id": "Baremetal",
+                "type": "ipv4_dhcp",
+                "link": "enp2s0",
+            }
+        ],
+        "services": [
+            {
+                "type": "dns",
+                "address": "8.8.8.8"
+            }
+        ]
+    }
+```
+
+This object will be reconciled by its own controller. When reconciled,
+the controller will add a label pointing to the Metal3Cluster that has nodes
+linking to this object. The spec contains a `metaData` and a `networkData` field
```

Comment:
What is the label for?

Reply:
This is a very good idea for the validation. We could also take an approach of letting the user define a map directly for the metadata, so it will be validated as a map[string]string. For the network, I also like the approach, though I would not bake in the functions, since the user might want to directly give the mac address for example. 

I am a bit unsure if we would take away some features from the network data. Considering the definition https://docs.openstack.org/nova/latest/_downloads/9119ca7ac90aa2990e762c08baea3a36/network_data.json there are different types of links, networks etc. Either we need to implement it so that we validate again all the fields, so re-implementing the nova specs, or we leave more room for the user, with the opportunity for mistakes

The label does not have a particular use, except for clarity.